plugins/README.adoc

```diff
@@ -7,3 +7,237 @@ NOTE: Work in Progress...
 * Review Ansible guidelines for https://docs.ansible.com/ansible/latest/dev_guide/developing_modules_best_practices.html[modules] and https://docs.ansible.com/ansible/latest/dev_guide/index.html[development].
 * Use https://pep8.org/[PEP8].
 * File headers and functions should have comments for their intent.
+
+
+== All plugin types should use argspec validation, inbound and outbound
+[%collapsible]
+===
+Explanations::
+
+Rationale::
+
+Examples::
+
+====
+
+== All plugin types should have documentation
+[%collapsible]
+====
+Explanations::
+
+Rationale::
+
+Examples::
+
+====
+
+== A plugin’s argspec should not be incrementally built but instead, be defined as a single object
```

Comment:
@cidrblock I'm questioning if we need to keep this recommendation given the later one:
> A complex plugin argspec should be derived from its documentation runtime if the plugin’s source is manually maintained.

If a complex argspec is not being automatically generated, I would disagree with this recommendation (there are cloud examples where trying to hand-write complex nested argspecs without splitting pieces out into vars is a huge pain).  I think generating complex specs is the better advice between then though.

What do you think?

Reply:
IMHO...

If the arspec is mantained by a human, I generally prefer:
1) That it be built runtime from the docstring, this ensures there is a single SOT for it (yeah, ansible-test does compare, but it still means double the updates)
2) If it can't be derived run time from the docstring (there are some limitations here and I've only seen it done to date in a limited fashion) I personally prefer a single data structure like this: https://github.com/ansible-collections/cisco.nxos/blob/main/plugins/module_utils/network/nxos/argspec/bgp_global/bgp_global.py rather than an incrementally built up data structure using a slew of `dict()` statements or even this incremental approach https://github.com/ansible-collections/amazon.cloud/blob/main/plugins/modules/backup_report_plan.py#L190

That said, both work, ( although I think I recall `dict()` being significantly slower than `{}`) and this is probably more of an individual dev preference.  I suspect some teams will decide on a pattern.

Maybe, "Coding styles for argspec definition should be consistent within a collection?"

Reply:
With deeply nested options dictated by a 3rd party API we've seen folks have difficulty with properly formatting the dictionaries and it creates a lot of frustrating back and forth in PRs and seriously ugly, difficult to review arg specs (AWS Auto Scaling comes to mind).  It's certainly a huge argument in favor of generating specs, but we have the code we have today.  :)

"Consistent within a collection" works for me.